engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt

```diff
@@ -44,28 +46,53 @@ internal class NadelHydrationValidation(
             error("Don't invoke hydration validation if there is no hydration silly")
         }
 
-        val errors = mutableListOf<NadelSchemaValidationError>()
-        for (hydration in hydrations) {
-            val actorService = services[hydration.serviceName]
-            if (actorService == null) {
-                errors.add(MissingHydrationActorService(parent, overallField, hydration))
-                continue
+        return hydrations
+            .asSequence()
+            .flatMap { hydration ->
+                val actorService = services[hydration.serviceName]
+                    ?: return@flatMap listOf(
+                        MissingHydrationActorService(parent, overallField, hydration),
+                    )
+
+                val actorServiceQueryType = actorService.underlyingSchema.queryType
+                val actorField = actorServiceQueryType.getFieldAt(hydration.pathToActorField)
+                    ?: return@flatMap listOf(
+                        MissingHydrationActorField(parent, overallField, hydration, actorServiceQueryType),
+                    )
+
+                getArgumentErrors(parent, overallField, hydration, actorServiceQueryType, actorField) +
+                    getOutputTypeIssues(parent, overallField, actorService, actorField) +
+                    getBatchHydrationErrors(parent, overallField, hydration, actorField)
             }
+            .toList()
+    }
+
+    private fun getBatchHydrationErrors(
+        parent: NadelServiceSchemaElement,
+        overallField: GraphQLFieldDefinition,
+        hydration: UnderlyingServiceHydration,
+        actorField: GraphQLFieldDefinition,
+    ): List<NadelSchemaValidationError> {
+        if (!hydration.isBatched && !actorField.type.unwrapNonNull().isList) {
+            return emptyList()
+        }
+
+        if (!actorField.type.unwrapNonNull().isList) {
+            // todo return an error here
+        }
 
-            val actorServiceQueryType = actorService.underlyingSchema.queryType
-            val actorField = actorServiceQueryType.getFieldAt(hydration.pathToActorField)
-            if (actorField == null) {
-                errors.add(MissingHydrationActorField(parent, overallField, hydration, actorServiceQueryType))
-                continue
+        val sourceArgCount = hydration.arguments
+            .count {
+                it.remoteArgumentSource.sourceType == OBJECT_FIELD
             }
 
-            val argumentIssues = getArgumentErrors(parent, overallField, hydration, actorServiceQueryType, actorField)
-            val outputTypeIssues = getOutputTypeIssues(parent, overallField, actorService, actorField)
-            errors.addAll(argumentIssues)
-            errors.addAll(outputTypeIssues)
+        if (sourceArgCount == 0) {
+            // todo return an error here
+        } else if (sourceArgCount > 1) {
+            // todo return an error here
         }
```

Comment:
There's some validation to be done around here, I will leave that to the other work stream :)